security/pkg/server/ca/server.go

```diff
@@ -153,29 +153,20 @@ func recordCertsExpiry(keyCertBundle *util.KeyCertBundle) {
 	if err != nil {
 		serverCaLog.Errorf("failed to extract root cert expiry timestamp (error %v)", err)
 	}
-	rootCertExpiryTimestamp.Record(rootCertExpiry)
-
-	rootCertPem, err := util.ParsePemEncodedCertificate(keyCertBundle.GetRootCertPem())
-	if err != nil {
-		serverCaLog.Errorf("failed to parse the root cert: %v", err)
-	}
-	rootCertExpirySeconds.ValueFrom(func() float64 { return time.Until(rootCertPem.NotAfter).Seconds() })
+	rootCertExpiryTimestamp.Record(float64(rootCertExpiry.Unix()))
+	rootCertExpirySeconds.ValueFrom(rootCertExpiry.Sub(time.Now()).Seconds)
 
 	if len(keyCertBundle.GetCertChainPem()) == 0 {
```

Comment:
Does this mean that if we transition to self-signed certs, the `certChainExpiry` metric will record the time until the last non-self-signed cert chain will expire? That doesn't seem right.

Reply:
@whitneygriffith

Reply:
We will have to investigate this further but conceptually I believe the `certChainExpiry` metric in the Istio project tracks the expiration time of the most recently updated certificate chain used by the cluster independent of whether its self signed or not. So it should reflect the expiration time of the new self-signed certificates. 

I think the answer is hidden in this code:

[`handleEvent`](https://github.com/istio/istio/blob/6aa63cb62372ba9e79940df6895273725765b5bb/pilot/pkg/bootstrap/istio_ca.go#L300)

Seems transitioning to a new root cert (self signed in your question), will require having both the old and new CA certs in the CA bundle and therefore both the [old and new cert chains](https://github.com/istio/istio/blob/6aa63cb62372ba9e79940df6895273725765b5bb/pilot/pkg/bootstrap/istio_ca.go#L337).  [`UpdateVerifiedKeyCertBundleFromFile`](https://github.com/istio/istio/blob/6aa63cb62372ba9e79940df6895273725765b5bb/pilot/pkg/bootstrap/istio_ca.go#L337) is a part of the transition process where [`VerifyAndSetAll`](https://github.com/istio/istio/blob/6aa63cb62372ba9e79940df6895273725765b5bb/security/pkg/pki/util/keycertbundle.go#L163) is called to set the new certificates, ensuring consistency and avoiding the use of old certificates.

Reply:
I may also be wrong. @jaellio knows best here!